## Getting data

First, we need to examine the data provided by the bank's credit department for the project. We start by importing `pandas` library for reading a csv-file into a DataFrame stored in `data` variable.

In [1]:
# Importing pandas library
import pandas as pd

Now, let's save the dataset contained in `bank_clients_data.csv` into a separate variable.

In [2]:
# Reading the dataset into a DataFrame
data = pd.read_csv('bank_clients_data.csv')

Print the first 20 rows of the DataFrame in order to see what the data looks like.

In [3]:
# Displaying the first 20 rows of the data
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

We can explore the data further by considering the summary information about the DataFrame.

In [4]:
# Displaying a concise summary
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Let's inspect what information the dataset contains:

1. The DataFrame has twelve columns, which have numeric and categorical values. 

2. `data` contains information about the clients of the bank with the following features:

    * `children` - number of children in a family;
    * `days_employed` - total number of days being employed;
    * `dob_years` - client's age;
    * `education` - client's education level;
    * `education_id` - education level indentifier;
    * `family_status` - client's family status;
    * `family_status_id` - family status identifier;
    * `gender` - client's gender;
    * `income_type` - type of employment;
    * `debt` - indicator of owing a debt to the bank;
    * `total_income` - monthly income;
    * `purpose` - purpose of taking out a loan.
3. The number of values across columns is not the same, signifying the fact that the data has missing values.

**Conclusions**

The dataset contains information about borrowers of the bank with numerous attributes, where each row identifies a separate debtor with specific data. The DataFrame is seen to have missing values present as well as potential data format problems, all of which warrant data preprocessing being carried out.

## Data preprocessing

### Missing values

Let's determine the number of missing values in the dataset.

In [5]:
# Displaying the number of missing values for each column
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

We see that there are two columns that contain missing values: `days_employed` and `total_income`. Deal first with the latter column. In order to get rid of missing values, we need to understand what the variable represents. `total_income` stores information about clients' incomes, meaning that the column possesses numeric values. Income is dependent upon the type of employment shown in `income_type` and therefore we can fill missing values with median income relative to a particular employment type.

In [6]:
# Filling in missing values with median
for inc_type in data['income_type'].unique():
    data.loc[(data['income_type'] == inc_type) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == inc_type), 'total_income'].median()

### Abnormalities

As can be seen from the data, we have abnormal values found in `days_employed` column: by no means can duration of employment be negative. Thus, we are dealing with abnormalities in the data. We can easily fix this by taking the absolute value of each data point in this column.

In [7]:
# Converting the data to a correct format
data['days_employed'] = data['days_employed'].abs()

Let's further explore this column and display median employment duration in relation to employment type.

In [8]:
# Displaying median values
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

We see that some income types, specifically, unemployed (`безработный`) and retired (`пенсионер`) have unusually high values. It is difficult to deal with this kind of abnormalities and since this column will be very informative to use in the subsequent analysis, we just leave the data as it is here.

Now, display the possible unique numbers of children that borrowers have.

In [9]:
# Printing unique values of children
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

It is clearly seen that there are two impossible values: 20 and -1. Let's get rid of them.

In [10]:
# Getting rid of abnormal values
data = data[(data['children'] != -1) & (data['children'] != 20)]
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Missing values (cont.)

Now we can fill in missing values in `days_employed` column with their median values depending on income type.

In [11]:
# Filling in missing values with median
for inc_type in data['income_type'].unique():
    data.loc[(data['income_type'] == inc_type) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == inc_type), 'days_employed'].median()

We can verify whether we have completely dealt with missing values issue.

In [12]:
# Checking changes made
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Data type changes

Let's change the data in `total_income` column from `float64` to `int64` for convenience. 

In [13]:
# Converting data type
data['total_income'] = data['total_income'].astype(int)

### Duplicates

We have some implicit duplicates in `education` column, specifically, using both capital and small letters to write down the same education level. Thus, let's convert all values in this column to strings with lowercase letters.

In [14]:
# Converting data to lowercase
data['education'] = data['education'].str.lower()

Now, we can take a look how many duplicates we have.

In [15]:
# Computing the number of explicit duplicates
data.duplicated().sum()

71

In [16]:
# Dropping all duplicates
data = data.drop_duplicates()

### Data categorization

To make further data analysis more convenient, let's create a categorical variable that will classify clients' incomes (in roubles) according to the following ranges:

* 0-30,000 => `'E'`
* 30,001-50,000 => `'D'`
* 50,001-200,000 => `'C'`
* 200,001-1,000,000 => `'B'`
* 1,000,001 and higher => `'A'`

For instance, a borrower with monthly income of, say, 200500 roubles will be put in `'B'` income category. We can program such a categorization by writing a function `categorize_income()` followed by `apply()` method.

In [17]:
# categorize_income() function
## A function takes one argument: total income (income)
## It then classifies it to a specific income category denoted by a letter "A-E"
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
# Applying the function to a column of the DataFrame
data['total_income_category'] = data['total_income'].apply(categorize_income)

Next, consider possible purposes that clients taking out loans may have.

In [19]:
# Printing unique purposes of taking a loan
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

We can see that it seems that there are lost of reasons for taking out loans in a bank, including real estate (`жилье`), buying a car (`автомобиль`), wedding (`свадьба`) and education (`образование`). But, in fact, all these purposes are simply written differently but actually refer to similar things. Hence, it is possible to also classify all these reasons into purpose categories:

* `'операции с автомобилем'` => buying a car
* `'операции с недвижимостью'` => real estate
* `'проведение свадьбы'` => wedding
* `'получение образования'` => education

In order to accomplish this, we can create a function `categorize_purpose()` and use a method `apply()` on it.

In [20]:
# categorize_purpose() function
## A function takes one argument: an element of a Series (row)
## It then classifies it according to its word structure to one of four categories
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
# Applying a function to a column of the DataFrame
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

**Conclusions**

We have managed to properly preprocess data by dealing with missing and abnormal values, changing the format of data and deleting duplicates, both expicit and implicit. 

Omissions in the source data were recorded in the columns `days_employed` and `total_income`. One possible reason for missing data may be the failure of an automated system for collecting information about borrowers. For example, when a borrower specifies information about himself in a form on the site and does not specify some information, the site may not notice this and simply send the information further (if the programmer forgot to make some cells mandatory). Another reason may be a purely human factor: if the borrower came to the bank and tells the bank employee information about himself, the employee may simply make a mistake in something and forget to specify the data, which will result in the appearance of `NaN` values.

Besides, we have managed to fill in missing values by using a median. As a rule, there are "outliers" in quantitative variables - values that are very different from other values in this column. The use of an average value can distort the data in this case and in order to avoid this, a median is used, which will take into account such sharply distinguished values and will not exaggerate the estimate of the missing value.

Next step is to actually start conducting creditworthiness analysis itself.

## Solvency determinants analysis

Now that the data has been preprocessed, we can get to analyzing possible factors that determine the ability of bank's clients to repay a loan on time.

### Number of children

In [22]:
# Computing the number of borrowers in each children group
data['children'].value_counts()

0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

To begin with, let's calculate the total number of borrowers by the number of children using the `value_counts()` method in the `children` column. It can be seen that the number of borrowers broken down by each group varies quite widely, which suggests that people with a large number of children do not take out loans so often.

In [23]:
# Aggregating data
children_grouped = data.groupby('children').agg({'debt': ['count', 'sum']})
# Computing number of borrowers who managed to repay on time by groups
children_grouped['no_debt'] = children_grouped['debt']['count'] - children_grouped['debt']['sum']
# Computing a fraction of borrowers who managed to repay on time by groups
children_grouped['no_debt_share'] = round(children_grouped['no_debt'] \
         / children_grouped['debt']['count'], 4)
# Sorting data
children_grouped.sort_values('no_debt_share', ascending=False)

debt       no_debt no_debt_share
          count   sum                      
children                                   
5             9     0       9        1.0000
0         14091  1063   13028        0.9246
3           330    27     303        0.9182
1          4808   444    4364        0.9077
2          2052   194    1858        0.9055
4            41     4      37        0.9024

Now let's try to answer the question posed. To do this, we aggregate the data using the `agg` method, counting the number of borrowers by groups using the `count` method, and using the `sum` method to count the number of people with debts (value of 1 in `debt` column refers to a person who has not paid on time). Next, we will count the number of people who have paid off the loan on time, finding the difference between these two columns.

Let's look at the data from the bank's point of view and use as a metric of loan repayment on time - "the proportion of people who have paid off a loan on time in a certain group" (`no_debt_share`). It can be seen that samples of borrowers with three or more children do not have many observations, especially for a group of people with five children. In other words, such samples are not representative enough and therefore it makes sense to compare samples with 0, 1 and 2 children.

In [24]:
# Printing a table for subsamples of borrowers with 0, 1, 2 children
children_grouped.loc[:2, :]

debt       no_debt no_debt_share
          count   sum                      
children                                   
0         14091  1063   13028        0.9246
1          4808   444    4364        0.9077
2          2052   194    1858        0.9055

Considering this grouping, it is obvious that with a larger number of children, the probability of repayment of the loan decreases: childless borrowers repay the loan on time in about 92.5% of cases, while for borrowers with a large number of children, this indicator is in the region of 90-91%. In other words, the repayment of the loan on time has an inverse relationship with the number of children of the borrower.

### Family status

In [25]:
# Computing the number of borrowers in each group by family status
data['family_status'].value_counts()

женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64

Let's apply the same strategy and look at the number of borrowers by marital status. Mostly loans are taken by people who are married (`женат / замужем`), and least of all by borrowers without official relations - widower/widow (`вдовец / вдова`) and divorced (`в разводе`). However, this does not mean that people who take out loans less often pay less on time.

In [26]:
# Aggregating data
family_grouped = data.groupby('family_status').agg({'debt': ['count', 'sum']})
# Computing the number of borrowers who repaid on time by groups
family_grouped['no_debt'] = family_grouped['debt']['count'] - family_grouped['debt']['sum']
# Computing the fraction of borrowers who repaid on time by groups
family_grouped['no_debt_share'] = round(family_grouped['no_debt'] \
        / family_grouped['debt']['count'], 4)
# Adding english translation of family statuses
family_grouped['family_status_eng'] = ['not married', 'divorced', 'widower/widow', 'civil marriage', 'married']
# Sorting data
family_grouped.sort_values('no_debt_share', ascending=False)

debt      no_debt no_debt_share family_status_eng
                       count  sum                                        
family_status                                                            
вдовец / вдова           951   63     888        0.9338     widower/widow
в разводе               1189   84    1105        0.9294          divorced
женат / замужем        12261  927   11334        0.9244           married
гражданский брак        4134  385    3749        0.9069    civil marriage
Не женат / не замужем   2796  273    2523        0.9024       not married

As in the previous point, the number of observations in each sample has a fairly strong influence on the results of the study, since a smaller sample will absorb less information, thus distorting them. Therefore, in this case it makes sense to consider categories that include, say, more than 2000 people. Thus, we will exclude the categories of widower/widows (`вдовец / вдова`) and divorced (`в разводе`) from the analysis and consider others.

In [27]:
rows = data['family_status'].unique()
family_grouped.loc[[rows[0], rows[1], rows[4]], :]

debt      no_debt no_debt_share family_status_eng
                       count  sum                                        
family_status                                                            
женат / замужем        12261  927   11334        0.9244           married
гражданский брак        4134  385    3749        0.9069    civil marriage
Не женат / не замужем   2796  273    2523        0.9024       not married

Now the dependence is more clearly traced: more family people are more responsible for timely repayment of debt, which is manifested in a high value of the share of refunds for such categories of borrowers. Thus, we have a direct relationship between the "formality" of family relations and the repayment of the loan on time.

### Total income

In [28]:
# Computing the number of borrowers in each group by income
data['total_income_category'].value_counts()

C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64

In [29]:
# Aggregating data
total_income_grouped = data.groupby('total_income_category').agg({'debt': ['count', 'sum']})
# Computing the number of people who repaid on time by groups
total_income_grouped['no_debt'] = total_income_grouped['debt']['count'] \
        - total_income_grouped['debt']['sum']
# Computing the fraction of people who repaid on time by groups
total_income_grouped['no_debt_share'] = round(total_income_grouped['no_debt'] \
        / total_income_grouped['debt']['count'], 4)
# Adding income ranges
total_income_grouped['income_range'] = ['>1,000,000', '200,001-1,000,000', '50,001-200,000', 
                                        '30,001-50,000', '0-30,000']
# Sorting the table
total_income_grouped.sort_values('no_debt_share', ascending=False)

debt       no_debt no_debt_share       income_range
                       count   sum                                         
total_income_category                                                      
D                        349    21     328        0.9398      30,001-50,000
B                       5014   354    4660        0.9294  200,001-1,000,000
A                         25     2      23        0.9200         >1,000,000
C                      15921  1353   14568        0.9150     50,001-200,000
E                         22     2      20        0.9091           0-30,000

The table shows a similar situation: the number of borrowers of categories A, E and D is extremely small, which may affect the results of the study. Borrowers in these groups are underrepresented and it would be better to exclude them from the analysis and consider the remaining categories.

In [30]:
rows = ['B', 'C']
total_income_grouped.loc[rows, :]

debt       no_debt no_debt_share       income_range
                       count   sum                                         
total_income_category                                                      
B                       5014   354    4660        0.9294  200,001-1,000,000
C                      15921  1353   14568        0.9150     50,001-200,000

So, we can see that the higher the borrower's income category, the higher the probability of debt repayment, so we have a direct relationship between these two variables.

### Loan purpose

In [31]:
# Computing the number of borrowers in each group by loan purpose
data['purpose_category'].value_counts()

операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64

In [32]:
# Aggregating data
purpose_grouped = data.groupby('purpose_category').agg({'debt': ['count', 'sum']})
# Computing the number of people who repaid on time by purpose
purpose_grouped['no_debt'] = purpose_grouped['debt']['count'] \
        - purpose_grouped['debt']['sum']
# Computing the fraction of people who repaid on time by purpose
purpose_grouped['no_debt_share'] = round(purpose_grouped['no_debt'] \
        / purpose_grouped['debt']['count'], 4)
# Adding english translation of loan purposes
purpose_grouped['purpose_category_eng'] = ['car', 'real estate', 'education', 'wedding']
# Sorting data
purpose_grouped.sort_values('no_debt_share', ascending=False)

debt      no_debt no_debt_share  \
                          count  sum                         
purpose_category                                             
операции с недвижимостью  10751  780    9971        0.9274   
проведение свадьбы         2313  183    2130        0.9209   
получение образования      3988  369    3619        0.9075   
операции с автомобилем     4279  400    3879        0.9065   

                         purpose_category_eng  
                                               
purpose_category                               
операции с недвижимостью          real estate  
проведение свадьбы                    wedding  
получение образования               education  
операции с автомобилем                    car

Basically, loans are taken for real estate transactions and, moreover, borrowers of this group more often repay loans, so this direction of lending should be considered by bank analysts.

**Conclusions**

By considering separate solvency-affecting factors, we have managed to identify possible profit-enhacing channels that a bank can focus on.

## Conclusion

So, in this study, the statistics of the solvency of customers of a credit institution were analyzed. To obtain the results, the data obtained were carefully preprocessed: duplicates were removed, abnormal and missing values were processed, and data categorization was carried out to simplify further data analysis.

Based on the results of the study, the following conclusion can be drawn. In order for the bank to increase its profits, the following groups of borrowers are the most profitable sources of funds for the bank:

* Childless;

* Married;

* People with salaries from 200,000 to 1,000,000 roubles;

* People taking out loans for real estate transactions.

In the future, it is recommendable that in the process of collecting data, a more careful approach to their processing is applied. In particular, it is advisable that missing values in the data be processed appropriately at the stage of sending them for analysis, or explanations be provided regarding the nature of their absence. The same applies to checking the correctness of the data format. This will lead to more efficient data analysis and the provision of more accurate recommendations and results.